src/audio/selector.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2019, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Lech Betlej <lech.betlej@linux.intel.com>
+ */
+
+/**
+ * \file audio/selector.h
+ * \brief Channel selector component header file
+ * \authors Lech Betlej <lech.betlej@linux.intel.com>
+ */
+
+#ifndef SELECTOR_H
+#define SELECTOR_H
+
+#include <stdint.h>
+#include <sof/audio/component.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/format.h>
+
+#if defined(__XCC__)
+#include <xtensa/config/core-isa.h>
+#endif
+
+/** \brief Selector trace function. */
+#define trace_selector(__e, ...)	trace_event(TRACE_CLASS_SELECTOR, __e, ##__VA_ARGS__)
+
+/** \brief Selector trace value function. */
+#define tracev_selector(__e, ...)	tracev_event(TRACE_CLASS_SELECTOR, __e, ##__VA_ARGS__)
+
+/** \brief Selector trace error function. */
+#define trace_selector_error(__e, ...)	trace_error(TRACE_CLASS_SELECTOR, __e, ##__VA_ARGS__)
+
+/**
+ * \brief Supported channel count on input
+ */
+#define SEL_IN_2CH 2
+#define SEL_IN_4CH 4
+
+/**
+ * \brief Supported channel count on input
+ */
+#define SEL_OUT_1CH 1
+#define SEL_OUT_2CH 2
+#define SEL_OUT_4CH 4
```

Comment:
can't we make this more generic and support selection of M channels in N, ideally with a mask?

Reply:
the component is written to enable certain use case in a finite time, from what I got from driver is that due to limitation in topology syntax there will be enum value passed from topology file to the component to select output channel, support for 2,4 channels in a pass through mode is added as a safety mechanism.  The component can be made more generic as a next step improvement.